File: mqttclients/dotnet/MQTTnet.Client.Extensions/MqttConnectionSettings.cs
using System.Diagnostics;
using System.Text;

namespace MQTTnet.Client.Extensions;

public enum AuthType
{
    X509,
    Basic
}

public class MqttConnectionSettings
{
    private const int Default_KeepAliveInSeconds = 30;
    private const string Default_CleanSession = "true";
    private const int Default_TcpPort = 8883;
    private const string Default_UseTls = "true";
    private const string Default_DisableCrl = "false";

    public string HostName { get; }
    public string? ClientId { get; set; }
    public string? CertFile { get; set; }
    public string? KeyFile { get; set; }
    public string? KeyFilePassword { get; set; }

    public AuthType Auth
    {
        get => !string.IsNullOrEmpty(CertFile) ? AuthType.X509 : AuthType.Basic;
    }
    public string? Username { get; set; }
    public string? Password { get; set; }
    public int KeepAliveInSeconds { get; set; }
    public bool CleanSession { get; set; }
    public int TcpPort { get; set; }
    public bool UseTls { get; set; }
    public string? CaFile { get; set; }
    public bool DisableCrl { get; set; }


    public MqttConnectionSettings(string hostname)
    {
        HostName = hostname;
        TcpPort = Default_TcpPort;
        KeepAliveInSeconds = Default_KeepAliveInSeconds;
        UseTls = Default_UseTls == "true";
        DisableCrl = Default_DisableCrl == "true";
        CleanSession = Default_CleanSession == "true";
    }

    public static MqttConnectionSettings FromConnectionString(string cs) => ParseConnectionString(cs);

    public static MqttConnectionSettings CreateFromEnvVars(string? envFile = "")
    {
        if (string.IsNullOrEmpty(envFile))
        {
            envFile = ".env";
        }

        if (File.Exists(envFile))
        {
            Trace.TraceInformation("Loading environment variables from {envFile}" + new FileInfo(envFile).FullName);
            foreach (string line in File.ReadAllLines(envFile))
            {
                string[] parts = line.Split('=', StringSplitOptions.RemoveEmptyEntries);
                if (parts.Length != 2)
                {
                    continue;
                }

                Environment.SetEnvironmentVariable(parts[0], parts[1]);
            }
        }
        else
        {
            Trace.TraceWarning($"EnvFile Not found in path {new DirectoryInfo(".").FullName} {envFile}");
        }
        static string ToUpperCaseFromPascalCase(string pascal) =>
            string.Concat(pascal.Select(x => Char.IsUpper(x) ? "_" + x : x.ToString())).ToUpper().TrimStart('_');

        static string Env(string name) =>
            Environment.GetEnvironmentVariable("MQTT_" + ToUpperCaseFromPascalCase(name)) ?? string.Empty;

        string hostname = Env(nameof(HostName));

        ArgumentException.ThrowIfNullOrEmpty(hostname, nameof(hostname));

        return new MqttConnectionSettings(hostname)
        {
            ClientId = Env(nameof(ClientId)),
            CertFile = Env(nameof(CertFile)),
            KeyFile = Env(nameof(KeyFile)),
            Username = Env(nameof(Username)),
            Password = Env(nameof(Password)),
            KeepAliveInSeconds = string.IsNullOrEmpty(Env(nameof(KeepAliveInSeconds))) ? Default_KeepAliveInSeconds : CheckForValidIntegerInput(nameof(KeepAliveInSeconds), Env(nameof(KeepAliveInSeconds))),
            CleanSession = Env(nameof(CleanSession)) == "true",
            TcpPort = string.IsNullOrEmpty(Env(nameof(TcpPort))) ? Default_TcpPort : CheckForValidIntegerInput(nameof(TcpPort), Env(nameof(TcpPort))),
            UseTls = string.IsNullOrEmpty(Env(nameof(UseTls))) ? true : CheckForValidBooleanInput(nameof(UseTls), Env(nameof(UseTls))),
            CaFile = Env(nameof(CaFile)),
            DisableCrl = Env(nameof(DisableCrl)) == "true",
            KeyFilePassword = Env(nameof(KeyFilePassword)),

        };
    }

    private static string GetStringValue(IDictionary<string, string> dict, string propertyName, string defaultValue = "")
    {
        string result = defaultValue;
        if (dict.TryGetValue(propertyName, out string? value))
        {
            result = value;
        }
        return result;
    }

    private static int GetPositiveIntValueOrDefault(IDictionary<string, string> dict, string propertyName, int defaultValue)
    {
        int result = defaultValue;
        if (dict.TryGetValue(propertyName, out string? stringValue))
        {
            if (int.TryParse(stringValue, out int intValue))
            {
                result = intValue;
            }
        }
        return result;
    }

    private static MqttConnectionSettings ParseConnectionString(string connectionString)
    {

        IDictionary<string, string> map = connectionString.ToDictionary(';', '=');
        string hostName = GetStringValue(map, nameof(HostName));
        ArgumentNullException.ThrowIfNull(hostName, nameof(hostName));

        MqttConnectionSettings cs = new(hostName)
        {
            ClientId = GetStringValue(map, nameof(ClientId)),
            KeyFile = GetStringValue(map, nameof(KeyFile)),
            CertFile = GetStringValue(map, nameof(CertFile)),
            Username = GetStringValue(map, nameof(Username)),
            Password = GetStringValue(map, nameof(Password)),
            KeepAliveInSeconds = GetPositiveIntValueOrDefault(map, nameof(KeepAliveInSeconds), Default_KeepAliveInSeconds),
            CleanSession = GetStringValue(map, nameof(CleanSession), Default_CleanSession) == "true",
            TcpPort = GetPositiveIntValueOrDefault(map, nameof(TcpPort), Default_TcpPort),
            UseTls = GetStringValue(map, nameof(UseTls), Default_UseTls) == "true",
            CaFile = GetStringValue(map, nameof(CaFile)),
            DisableCrl = GetStringValue(map, nameof(DisableCrl), Default_DisableCrl) == "true"
        };
        return cs;
    }

    private static int CheckForValidIntegerInput(string envVarName, string envVarValue)
    {
        if (int.TryParse(envVarValue, out var result))
        {
            return result;
        }

        throw new ArgumentException($"An environment variable's type was specified incorrectly: {envVarName}={envVarValue}. Expecting an integer value.");
    }

    private static bool CheckForValidBooleanInput(string envVarName, string envVarValue)
    {
        if (bool.TryParse(envVarValue, out var result))
        {
            return result;
        }

        throw new ArgumentException($"An environment variable's type was specified incorrectly: {envVarName}={envVarValue}. Expecting a boolean value.");
    }

    private static void AppendIfNotEmpty(StringBuilder sb, string name, string val)
    {
        if (!string.IsNullOrEmpty(val))
        {
            sb.Append($"{name}={val};");
        }
    }

    public override string ToString()
    {
        StringBuilder result = new ();
        AppendIfNotEmpty(result, nameof(HostName), HostName!);
        AppendIfNotEmpty(result, nameof(TcpPort), TcpPort.ToString());
        AppendIfNotEmpty(result, nameof(Username), Username!);
        AppendIfNotEmpty(result, nameof(CleanSession), CleanSession.ToString());
        AppendIfNotEmpty(result, nameof(KeepAliveInSeconds), KeepAliveInSeconds.ToString());
        AppendIfNotEmpty(result, nameof(CertFile), CertFile!);
        AppendIfNotEmpty(result, nameof(KeyFile), KeyFile!);
        AppendIfNotEmpty(result, nameof(CaFile), CaFile!);
        AppendIfNotEmpty(result, nameof(ClientId), ClientId!);
        AppendIfNotEmpty(result, nameof(UseTls), UseTls.ToString());
        AppendIfNotEmpty(result, nameof(Auth), Auth!.ToString());
        result.Remove(result.Length - 1, 1);
        return result.ToString();
    }
}

File: mqttclients/ts/mqttjsClientExtensions/src/mqttConnectionSettings.ts
import { config } from 'dotenv';

enum AuthType {
    X509 = 'X509',
    Basic = 'Basic'
}

const defaultHostname = 'localhost';
const defaultKeepAliveInSeconds = 30;
const defaultCleanSession = true;
const defaultTcpPort = 8883;
const defaultUseTls = true;
const defaultDisableCrl = false;

export class MqttConnectionSettings {
    private _hostname = defaultHostname;
    public clientId = '';
    public certFile = '';
    public keyFile = '';
    public keyFilePassword = '';
    public username = '';
    public password = '';
    public keepAliveInSeconds: number = defaultKeepAliveInSeconds;
    public cleanSession: boolean = defaultCleanSession;
    public tcpPort: number = defaultTcpPort;
    public useTls: boolean = defaultUseTls;
    public caFile = '';
    public disableCrl: boolean = defaultDisableCrl;

    constructor(hostname: string) {
        this._hostname = hostname;
    }

    public static fromConnectionString(connectionString: string): MqttConnectionSettings {
        return MqttConnectionSettings.parseConnectionString(connectionString);
    }

    public static createFromEnvVars(envFilePath = '.env'): MqttConnectionSettings {
        // Load environment variables from .env file using the dotenv package
        const envConfig = config({ path: envFilePath });

        if (envConfig.error) {
            throw new Error(envConfig.error.message);
        }

        const hostname = envConfig.parsed?.MQTT_HOST_NAME;

        if (!hostname) {
            throw new Error('MQTT_HOST_NAME environment variable is not set');
        }

        if (envConfig.parsed?.MQTT_PASSWORD && !envConfig.parsed?.MQTT_USERNAME) {
            throw new Error('MQTT_USERNAME environment variable is required if MQTT_PASSWORD is set');
        }

        const cs = new MqttConnectionSettings(hostname);
        cs.tcpPort = Number(envConfig.parsed?.MQTT_TCP_PORT ?? defaultTcpPort);
        cs.useTls = Boolean(envConfig.parsed?.MQTT_USE_TLS === undefined ? defaultUseTls : envConfig.parsed?.MQTT_USE_TLS);
        cs.cleanSession = Boolean(envConfig.parsed?.MQTT_CLEAN_SESSION === undefined ? defaultCleanSession : envConfig.parsed?.MQTT_CLEAN_SESSION);
        cs.keepAliveInSeconds = Number(envConfig.parsed?.MQTT_KEEP_ALIVE_IN_SECONDS ?? defaultKeepAliveInSeconds);
        cs.clientId = envConfig.parsed?.MQTT_CLIENT_ID ?? '';
        cs.username = envConfig.parsed?.MQTT_USERNAME ?? '';
        cs.password = envConfig.parsed?.MQTT_PASSWORD ?? '';
        cs.certFile = envConfig.parsed?.MQTT_CERT_FILE ?? '';
        cs.keyFile = envConfig.parsed?.MQTT_KEY_FILE ?? '';
        cs.caFile = envConfig.parsed?.MQTT_CA_FILE ?? '';

        return cs;
    }

    public static parseConnectionString(connectionString: string): MqttConnectionSettings {
        let cs: MqttConnectionSettings;

        try {
            const csElements = connectionString.split(';');
            const csMap = new Map<string, string>();
            for (const element of csElements) {
                const [key, value] = element.split('=');
                csMap.set(key, value);
            }
            const csObj = Object.fromEntries(csMap);

            cs = new MqttConnectionSettings(csObj.HostName);
            cs.clientId = csObj?.ClientId ?? '';
            cs.keyFile = csObj?.KeyFile ?? '';
            cs.certFile = csObj?.CertFile ?? '';
            cs.username = csObj?.Username ?? '';
            cs.password = csObj?.Password ?? '';
            cs.keepAliveInSeconds = Number(csObj?.KeepAliveInSeconds ?? defaultKeepAliveInSeconds);
            cs.cleanSession = Boolean(csObj?.CleanSession === undefined ? defaultCleanSession : csObj?.CleanSession);
            cs.tcpPort = Number(csObj?.TcpPort ?? defaultTcpPort);
            cs.useTls = Boolean(csObj?.UseTls === undefined ? defaultUseTls : csObj?.UseTls);
            cs.caFile = csObj?.CaFile ?? '';
            cs.disableCrl = Boolean(csObj?.DisableCrl === undefined ? defaultDisableCrl : csObj?.DisableCrl);
        }
        catch (ex) {
            throw new Error(`Error while parsing connection string: ${ex.message}`);
        }

        return cs;
    }

    public get hostname(): string {
        return this._hostname;
    }

    public get auth(): AuthType {
        return this.certFile ? AuthType.X509 : AuthType.Basic;
    }
}
